builtin/logical/ssh/util.go

```diff
@@ -0,0 +1,206 @@
+package ssh
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"fmt"
+	"io"
+	"net"
+	"strings"
+
+	"github.com/hashicorp/vault/logical"
+
+	"golang.org/x/crypto/ssh"
+)
+
+// Transfers the file  to the target machine by establishing an SSH
+// session with the target. Uses the public key authentication method
+// and hence the parameter 'key' takes in the private key. The fileName
+// parameter takes an absolute path.
+func uploadPublicKeyScp(publicKey, username, ip, port, key string) error {
+	dynamicPublicKeyFileName := fmt.Sprintf("vault_ssh_%s_%s.pub", username, ip)
+	session, err := createSSHPublicKeysSession(username, ip, port, key)
+	if err != nil {
+		return err
+	}
+	if session == nil {
+		return fmt.Errorf("invalid session object")
+	}
+	defer session.Close()
+	go func() {
```

Comment:
This looks like a race condition to me, we are running the upload concurrent with the session run.